tetraelf-hierarchy/hierarchy-portal/impl/src/java/org/sakaiproject/portal/impl/HierarchySiteNeighbourhoodService.java

```diff
@@ -191,17 +193,51 @@ public String parseSiteAlias(String alias) {
 	}
 
 	@Override
-	public String getRedirect(String siteId) {
-		if (siteId != null) {
-			PortalNode node = portalHierarchyService.getNodeByUrl(siteId);
-			if (node instanceof PortalNodeRedirect) {
-				return ((PortalNodeRedirect)node).getUrl();
-			}
+	public Redirect getRedirect(String siteId) {
+		if (StringUtils.isNotBlank(siteId)) {
+			Redirect redirect = new Redirect();
 			// If we have trailing seperators pull them off
 			Matcher matcher = trimTailingSeparators.matcher(siteId);
 			if (matcher.matches()) {
 				// This is a relative URL that the container will make absolute
-				return serverConfigurationService.getString("portalPath")+ "/"+ "site"+ "/"+ matcher.group(1);
+				redirect.setUrl(serverConfigurationService.getString("portalPath")+ "/"+ "site"+ "/"+ matcher.group(1));
+				return redirect;
+			}
+			PortalNode node = null;
+			String trimmedPath = "";
+			int length = StringUtils.countMatches(siteId, PortalHierarchyService.SEPARATOR);
+			for(int i = length; i >= 1 ; i--){
+				node = portalHierarchyService.getNodeByUrl(siteId);
+				//if node null, check for the redirect of subsites
+				if(node == null){
+					//get the index of last occurrence of Separator in the path
+					int index = siteId.lastIndexOf(PortalHierarchyService.SEPARATOR);
+					if(index > 0){
+						String firstPart = siteId.substring(0, index);
+						String tempString = PortalHierarchyService.SEPARATOR + siteId.substring(index + 1);
+						//save the trimmed path if there is a subsite redirect
+						trimmedPath = (StringUtils.isBlank(trimmedPath)) ? tempString : (tempString + trimmedPath) ;
+						siteId = firstPart;
```

Comment:
Generally bad style to re-assign parameters to methods. Better to create a new variable and set it to the parameter at the start of the loop.